students/navdeep/session04/mailroom_part2.py

```diff
@@ -0,0 +1,147 @@
+
+def displayMenu():
+    """
+    displays the menu of options to the user
+    :return:
+    """
+    print()
+    str_menu = "1. Type 't' to Send a Thank You to add a name, donation and thank you card\n"  \
+               "2. Type 'r' to Create a Report and see a list of the donors\n" \
+               "3. Type 's' to Send letters to everyone\n" \
+               "4. Type 'e' to exit the program"
+    print(str_menu)
+    print()
+
+def addNewDonation(donor_list):
+    """
+    Allows user to add a new donor and their donation to the list
+    :param donor_list: the list of donors and their donations
+    """
+    new_donor = input("1. Enter the name of a new donor\n"
+                      "2. Type 'list' to list the donor names\n"
+                      "3. Type 'exit' to return to main menu\n"
+                      "Your Selection:  ")
+    print()
+    if new_donor.lower() == 'list':
+        showDonorNames(donor_list)
+    elif new_donor.lower() == 'exit':
+        return
+    else:
+        new_donation = float(input("Enter a new donation amount: "))
+
+        while new_donation < 1:
+            new_donation = float(input("Enter something greater than 0: "))
+        if new_donor.strip() in donor_list:
+            donor_list[new_donor].append(new_donation)
+        else:
+            donor_list[new_donor.strip()] = [new_donation]
+        printThankYou(new_donor, new_donation)
+
+def showDonorNames(donor_list):
+    """
+    will show the unique names of donors
+    :param donor_list: the list of donors and their donations
+    """
+    print("Here are the list of donors: ")
+    for donor in donor_list.keys():
+        print(donor)
+
+def printThankYou(newest_donor, newest_donation):
+    """
+    Prints out a thank you message after someone has donated
+    :param newest_donor: the name of the newest donor
+    :param newest_donation: the amount the new donor donated
+    """
+    thank_you_message = "Thank you {} for you generous donation " \
+                        "of ${:.2f}.".format(newest_donor.title(), round(newest_donation,2))
+    print(thank_you_message)
+    print()
+
+def reportHeader():
+    """
+    The header of the report
+    """
+    print("{:25s} | {:11s} | {:9s} | {:12s}".format(
+        "Donor Name", "Total Given", "Num Gifts", "Average Gift"))
+    print("-" * 66)
+
+def createReport(donor_list):
+    """
+    Creates a table of the report that shows the donor names, how many times they donated
+    the average donation and total donations
+    :param donor_list: the list of donors and their donation amounts
+    """
+    total_give = 0.00
+    num_gifts = 0
+    avg_gift = 0
+    report_dict = {}
+    for donor, donations in donor_list.items():
+        num_gifts = len(donations)
+        total_give = sum(donations)
+        avg_gift = total_give / num_gifts
+        report_dict[donor] = [round(total_give,2), num_gifts, round(avg_gift, 2)]
+    reportHeader()
```

Comment:
This is great. Consider making a separate function for all of the printing, not just the header, then it will be easier to test later. :)